### DISCUSSION TITLE
Клиент серверное взаимодействие

### DISCUSSION THREAD

**Discussion by pingvin1308** [cleannetcode/CsharpForBeginers#20] [Новички]
https://github.com/emedvedeva/CsharpForBeginners/tree/main/Shop
https://github.com/emedvedeva/CsharpForBeginners/blob/main/HomeTask5/Program.cs
https://github.com/coder1coder/Shop
https://github.com/trulander/CsharpEducation/tree/main/ShowCase
https://github.com/Z1pP/CsharpForBeginners/tree/master/Shop

Доработать магазин, добавив работу через клиент, сервер.

+ Добавить применение делегатов в клиент и сервер. Например сделать нотификацию об ошибках.

```
internal delegate void Notify(string message);

Notify notify = message => Console.WriteLine(message);
```

Более подробные примеры можно посмотреть тут: https://github.com/cleannetcode/CsharpForBeginners/tree/develop/Delegates

https://t.me/joinchat/ETnMMXnhJng4YmJi - так же можно писать и задавать вопросы тут :)

**Comment by trulander** [comment]
Привет. Моя домашка готова, я форкнул проект coder1coder и доработал его. 
https://github.com/trulander/Shop

Не хотел сильно усложнять, но идея у меня была в том что клиент может подключиться к серверу как будтое ко второй консоле, в то время как серверное приложение имеет свою независимую консоль, но так как coder1coder не реализовал хотя бы мнимую реализацию БД сейчас 2 консоли независимы друг от друга и данные по товарам\ветринам и т.д также независимы.

Я немного схитрил и использовал "запрещенные" приемы, мне показалось разобраться с теми методами которые я использовал было проще чем написать приложение без них.

**Reply by trulander** [reply]
и да, coder1coder что-то там намудрил и поломал программу, не все функции работают как задумывалось, я это заметил только когда уже завершил свою переделку. Неделю назад у него вроде бы все работало, кроме корзины:-)

**Reply by coder1coder** [reply]
> .. но так как coder1coder не реализовал хотя бы мнимую реализацию БД ..

ну не знаю, не знаю, почему тебе не хватило объектов с классами реализующими IRepository? Достаточно для хранения в памяти

> и да, coder1coder что-то там намудрил и поломал программу, не все функции работают как задумывалось, я это заметил только когда уже завершил свою переделку. Неделю назад у него вроде бы все работало, кроме корзины:-)

посмотри по коммитам, там все изменения

**Reply by coder1coder** [reply]
Может быть ты не все зависимости собрал и подключил? :)

**Reply by trulander** [reply]
> ну не знаю, не знаю, почему тебе не хватило объектов с классами реализующими IRepository? Достаточно для хранения в памяти
у тебя каждый класс реализует свое хранилище, и нет общего хранилища. Я не хотел переписывать твою реализацию для централизованного хранения данных, просто изначальной задачи не было на это.

>посмотри по коммитам, там все изменения
>Может быть ты не все зависимости собрал и подключил? :)
Я заметил что не все работает, к примеру удаление витрин и товаров, тогда когда делал финальное тестирование и отладку своего кода. Я сначала подумал что у меня что то не работает, потом запустил твой код из твоего репозитория и там было точно такое же поведение. Исправлять я уже ничего не хотел, но так то конечно можно было и покопаться,е но решил остановиться своей реализацией, пока)
И если бы у меня что-то не собралось, у меня бы программа наверное не запустилась твоя.

**Reply by coder1coder** [reply]
Условие необходимо изменить на id < 1

[для товаров тут](https://github.com/coder1coder/Shop/blob/0ce51b59fd4bb9cfb5fe04f49137f5fb99efdbd6/Shop/Controller/ShopController.cs#L170-L172)

[для витрин тут](https://github.com/coder1coder/Shop/blob/0ce51b59fd4bb9cfb5fe04f49137f5fb99efdbd6/Shop/Controller/ShopController.cs#L276-L278)

запушил багфиксы

**Reply by pingvin1308** [reply]
Молодцы, горжусь! :)

**Comment by trulander** [comment]
Всем привет. Сделал обновленную домашку, на этот раз уже делал на своей программе уже с делегатами и уведомлениями
https://github.com/trulander/CsharpEducation/tree/main/ShowCase
Доработал свою программу до клиент серверного взаимодействия + добавил делегаты для локальных уведомлений об действиях со стороны пользователя локальной части программы и серверной и через эти же делегаты происходит обновление локального интерфейса если с клиентского приложения(серверной части основного приложения) вносились изменения.

Я добавил много комментариев по коду и по методам, думаю теперь логика работы приложения должна пониматься легче.

Изначально может показаться что у клиентского приложения и основного одна и таже консоль, но это не так, они просто синхронизируются так как база данных у них общая и действия можно производить из обоих консолей независимо, а результат будет виден везде.

**Reply by pingvin1308** [reply]
Любопытная работа. Есть над чем поработать)
Вижу, что ты уже достаточно хорошо разобрался в основах.
Скоро мы будем работать над структурами программ и учиться их делать грамотно, тогда у тебя получится сделать код в разы лучше. Сейчас просто акцентируй внимание на удобстве восприятия кода сторонним разработчиком)
Думаю о других и делай код как можно более открытым к изменениям. Но это все общие рекомендации.

**Comment by 8bit-com** [comment]
https://github.com/8bit-com/NetMarket.git чужой код не осилил сделал свой магазин, не весь функционал реализован, но по сети работает. Ума не приложу как сюда интерфейсы запихать и вообще зачем они здесь. И имена функций я не умею придумывать

**Reply by pingvin1308** [reply]
Ничего со временем придет, главное практиковать. В целом неплохая простенька программа получилась. Скоро будет как раз переключаться на изучение asp.net. Там будет поднимать заодно теорию, связанную с разделением проектов. Тут нам часто будут выручать интфейсы

**Reply by pingvin1308** [reply]
Также рекомендую глянуть реализации других участников)

**Comment by blood-mif** [comment]
https://github.com/blood-mif/CsharpForBeginners-2
Форкнул Z1pP , описал в readme client

**Reply by pingvin1308** [reply]
Я что то не нашел реализации сервера. Клиент выглядит вполне неплохо. Хорошо бы посмотреть как ты сделал сервер и можно будет начать изучать фреймворки)

**Reply by blood-mif** [reply]
https://github.com/blood-mif/CsharpForBeginners-2/blob/master/Shop/StartProgram.cs
68-168 строка, если я правильно понял вопрос=)

**Comment by emedvedeva** [comment]
Моя реализация клиент-серверного взаимодействия. Я взяла проект Z1pP.
https://github.com/emedvedeva/CsharpForBeginners-1/tree/develop
Реализовала только витрины, работать с продуктами мой проект не умеет.
1. Добавила класс ShowcaseList в модели
2. Не работало редактирование витрины из-за неправильной проверки на размер. Поправила
3. Удалила метод AddDefaultProduct() - он был пустой
4. У витрины много названий - запуталась - ShowCase, ProductStore, ShowWindow, но всех их оставила
5. Проверки на валидность вводимых значений пользователем при создании витрины не добавляла. Оставила как было сделано в проекте Z1.

**Reply by pingvin1308** [reply]
Хорошая работа, теперь можно смело думать об освоении asp.net core ;)

**Reply by Mopkovka13** [reply]
Твой проект очень помог разобраться, а-то сам я закопался ;)
